File: structy/structy-js/Binary-Tree/treeMinValue.js
// Write a function, treeMinValue, that takes in the root of a binary tree that contains number values. The function should return the minimum value within the tree.
// You may assume that the input tree is non-empty.

class Node {
	constructor(val) {
		this.val = val;
		this.left = null;
		this.right = null;
	}
}

//!breadth first
const treeMinValue = (root) => {
	const queue = [root]; //set queue to root node
	let min = root.val;
	while (queue.length > 0) {
		const node = queue.shift();
		if (node.val < min) {
			min = node.val;
		}
		if (node.left !== null) queue.push(node.left);
		if (node.right !== null) queue.push(node.right);
	}
	return min;
};

//! recursive
// const treeMinValue = (root) => {
//     if (root === null) return Infinity;
//     const smallestLeftValue = treeMinValue(root.left);
//     const smallestRightValue = treeMinValue(root.right);
//     return Math.min(root.val, smallestLeftValue, smallestRightValue);
//   };

//! depth first
// const treeMinValue = (root) => {
//     const stack = [root];

//     let smallest = Infinity;
//     while (stack.length) {
//       const current = stack.pop();
//       if (current.val < smallest) smallest = current.val;

//       if (current.left !== null) stack.push(current.left);
//       if (current.right !== null) stack.push(current.right);
//     }
//     return smallest;
//   };

//! TEST__00
const a = new Node(3);
const b = new Node(11);
const c = new Node(4);
const d = new Node(4);
const e = new Node(-2);
const f = new Node(1);

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//       3
//    /    \
//   11     4
//  / \      \
// 4   -2     1

console.log(treeMinValue(a)); // -> -2

//! TEST__01
// const a = new Node(5);
// const b = new Node(11);
// const c = new Node(3);
// const d = new Node(4);
// const e = new Node(14);
// const f = new Node(12);

// a.left = b;
// a.right = c;
// b.left = d;
// b.right = e;
// c.right = f;

// //       5
// //    /    \
// //   11     3
// //  / \      \
// // 4   14     12

// treeMinValue(a); // -> 3

//! TEST__02
// const a = new Node(-1);
// const b = new Node(-6);
// const c = new Node(-5);
// const d = new Node(-3);
// const e = new Node(-4);
// const f = new Node(-13);
// const g = new Node(-2);
// const h = new Node(-2);

// a.left = b;
// a.right = c;
// b.left = d;
// b.right = e;
// c.right = f;
// e.left = g;
// f.right = h;

// //        -1
// //      /   \
// //    -6    -5
// //   /  \     \
// // -3   -4   -13
// //     /       \
// //    -2       -2

// treeMinValue(a); // -> -13

//! TEST__03
// const a = new Node(42);

//    42

// treeMinValue(a); // -> 42

File: structy/structy-py/3-Binary-Tree/treeMinValue.py
# Write a function, tree_min_value, that takes in the root of a binary tree that contains number values. The function should return the minimum value within the tree.

# You may assume that the input tree is non-empty.

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

# #! iterative
# def tree_min_value(root):
#     if root is None:
#         return None
#     min_val = float('inf')
#     stack = [root]
#     while stack:
#         node = stack.pop()
#         if node.val < min_val:
#             min_val = node.val

#         if node.left is not None:
#             stack.append(node.left)
#         if node.right is not None:
#             stack.append(node.right)
#     return min_val

#! recursive


def tree_min_value(root):
    if root is None:
        return float('inf')
    left = tree_min_value(root.left)
    right = tree_min_value(root.right)
    return min(root.val, left, right)


#!TEST_00
a = Node(3)
b = Node(11)
c = Node(4)
d = Node(4)
e = Node(-2)
f = Node(1)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#       3
#    /    \
#   11     4
#  / \      \
# 4   -2     1
print(tree_min_value(a))  # -> -2

#!TEST_01
a = Node(5)
b = Node(11)
c = Node(3)
d = Node(4)
e = Node(14)
f = Node(12)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#       5
#    /    \
#   11     3
#  / \      \
# 4   14     12

print(tree_min_value(a))  # -> 3

#!TEST_02
a = Node(-1)
b = Node(-6)
c = Node(-5)
d = Node(-3)
e = Node(-4)
f = Node(-13)
g = Node(-2)
h = Node(-2)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f
e.left = g
f.right = h

#        -1
#      /   \
#    -6    -5
#   /  \     \
# -3   -4   -13
#     /       \
#    -2       -2

print(tree_min_value(a))  # -> -13

#!TEST_03
a = Node(42)

#        42

print(tree_min_value(a))  # -> 42
